### ISSUE TITLE
lws pipe = true doens't response when sending http request with high frenquency

### ISSUE THREAD

**Issue by qdztxc** [warmcat/libwebsockets#2499]
Hi Andy,

I'm using lws as http client. The server is third parth. The network between client and server is super good.
The server may refuse the request without response but I got info from you that lws has internal timer to handle this. 

I need to send thoudands of requets at very short time. I try to do test with 100 requests every 1ms.
In pseudocode：
while (--total_test_times > 0)
    for (int i = 0; i < 100; ++i)
        send_http_request
    sleep 1ms

fd_limit_per_thread is set to 1048578. max_http_header_pool = fd_limit_per_thread.

When pipe = false, though it's slow but it seems work well at last. 35598 of 35600 requests get reponse.
When pipe = true, it doesn't work. 2424 of 26860 requests get response.
Do you konw how to resolve this? 
Since pipe = true has better performance under low pressure I don't plan to change pipe to false by default.

Thanks and regards.

**Comment by lws-team** [comment]
> sleep 1ms

That does do something, but probably not what you expect if it blocks the event loop from operating.

You're describing what you do, but there is nothing about how lws feels about things from the logs.  So it's not surprising you don't know why you get these results.

What do the logs say happened with the 92% that failed in the "pipe=true" (whatever that means... enabled pipelining in the client creation info?) case?  Timeouts?  The connection is broken into several async steps including DNS lookup, tcp connect, tls establishment if you have it, and then states related to the http transaction.  It's possible with the avalanche of new connections, it is unable to get back to the first connections before they timeout due to having to sit through starting tens of thousands of other ones first.

Look at the logs...

**Comment by qdztxc** [comment]
I couldn't find obvious hint from log. :(
I tried to dump log to a file. This time 2595/28170 requests get response.
I attached the log file. Would you please help to have a look? 

Thanks and regards

[http_flush_lws.log](https://github.com/warmcat/libwebsockets/files/7579728/http_flush_lws.log)

**Comment by lws-team** [comment]
You can increase the log verbosity, but I can see the later wsi have been waiting >20s, they are likely to be timing out.

Look into context creation info `timeout_secs` and `connect_timeout_secs`

**Comment by qdztxc** [comment]
timeout_secs = 0
connect_timeout_secs = 360

I change log level to 
      	lws_set_log_level(LLL_USER | LLL_ERR | LLL_WARN | LLL_NOTICE | LLL_INFO | LLL_DEBUG, NULL);
and attached new log. This time 808/36120 succeeded.
Would you please help to have a look?

Thanks and regards

[http_flush_lws.zip](https://github.com/warmcat/libwebsockets/files/7588783/http_flush_lws.zip)

**Comment by lws-team** [comment]
Mmm thanks, but I don't really want to be inside your debugging loop looking at hundreds of megs of logs as if it's my problem to "make it work".

Why are you creating thousands of connections at once like this?  If you pipeline them, you are just clogging the thing up with queued connections before they can be used.  It seems kind of crazy.  If you want better advice, explain more clearly what you really need to do.

If you do have some compelling reason to do this, why not manage how many connections you have in flight by creating additional ones as you see that current ones have completed?  It's not going to be any slower overall.

**Comment by qdztxc** [comment]
I'm doing high frequency crypto trading which uses websocket for marketdata and http for send/cancel orders. I'm using lws for both websocket and http.

Usually a system of HFT is event driven.  If there is a signal triggered, thousands of orders (https) may be sent simultaneously. So is order cancellation.
When the signal triggers depends on market conditions. It doesn't mean I need to send thousands of http request all the time. But when there is good chance to trade, I need to send so many requests as soon as possible and want to get response (no matter succeed or fail or timeout) at the first opportunity.

“If you do have some compelling reason to do this, why not manage how many connections you have in flight by creating additional ones as you see that current ones have completed? It's not going to be any slower overall.”
——
For our case I need to send the these orders out in very short time after signal triggers. The network delay between client (our system) and server (crypto exchange) is several milliseconds which is quite slow from HFT point of view. Our system use dozens of  micro seconds from process market data to generate signal and then send orders to lws. 

Thanks and regards

**Comment by lws-team** [comment]
So the current naive scheme you have of just creating 1000s of connections that are slowly attempted over 20s sounds like it won't do what you want at all.

At any time new market data may come that counterindicates what you were desperate to spam out a few moments ago, it may be that still continuing to request what looked like a great idea just now has become disasterous.  So you need a much smaller transaction queue managed more dynamically that choking everything with a massive spam that takes time to go through.

You're doing something extreme, the default single instance of a singlethreaded event loop will not scale to what you want, you need to parallelize to increase the amount of transactions per second to the peer.  Parallelizing inside one box will probably do but it is better if it doesn't preclude multiple boxes.

And, not all latency, even on the hot path, is bad, if eating it buys you overall reduced latency downstream.

Putting all that together, what I would do is change your trading decision part to end with a local ws server listening on a unix domain socket or tcp.  Other separate processes (on same or different boxes if tcp) connect to that, and provide 'onward trading services' as handed out by the trading decision part over ws, eg as JSON.  The 'onward' processes report their progress back over the same ws link, and can respond to new commands dynamically.

This scales much better, since n 'onward' processes can join the trading ws server, n times the connections to the remote service are being handled in parallel (ie, each process has its on lws_context going on its own thread).  Onward processes can join and leave the ws server dynamicallly botnet style, if it listens on tcp, increasing the number of IPs that are doing the onward connections.

These 'onward' processes each just have a handful of queued requests in flight and don't exceed the transaction budget or TTL handed out to it by the transaction decision part.

**Comment by qdztxc** [comment]
Thanks for your advice.
Our system is hard to split into different processes. Does one process with multi worker threads (each work thread with individual lws_context) work?

Besides in this extreme condition, is it better to replace poll with epoll?

I'm still wondering no mater how slow it is almost all requrests get response when pipe = false. But only few responses receives when pipe = true.
I dont' expect all requests will succeed but at least they could report the failure.

Last but not least, is it reasonable to set timeout_secs = 0 and connect_timeout_secs = 360 in my scenario?

Thanks and regards.

**Comment by lws-team** [comment]
> Our system is hard to split into different processes.

If you use JSON to communicate with these onward processes, it's not going to be that hard and you end up with a faster, much better and more scaleable architecture.

> Does one process with multi worker threads (each work thread with individual lws_context) work?

Yes, but it means you can't extend the onward part to work from distributed routes / IPs over different boxes.  That seems like it should be interesting for a few reasons.  You also tend to get better quality code working inbetween process boundaries.

> is it better to replace poll with epoll?

You're welcome to do whatever you think best, it's FOSS.  Or you can take some clearly reasoned advice that actually addresses your problem rather than move the deckchairs on the Titanic.

> But only few responses receives when pipe = true.

Pipelining means transactions are accepted and queued until it's their turn.  Only when it is their turn do they actually go to the peer, so with what you are doing, they spend 20s at the client waiting for their turn and timeout before they can even do anything, it's a naive and ineffectual way.

> timeout_secs = 0

This is the same as 20s, since the way you're approaching it at the moment the transactions wait longer than 20s, this won't change your problem.  The bigger problem is how you're trying to do it at all, because if you increase the timeout to eg, 300s, you are saying that you want to perform transactions decided from conditions 5m ago.  Inbetween talking about how 100us is too much latency... the whole approach is wrong.
